httputil/errors.go

```diff
@@ -37,27 +44,43 @@ func ErrorHandler(w http.ResponseWriter, r *http.Request, err error) {
 		return
 	}
 
-	logctx.AddToContext(r.Context(), LogErrorMessage, err)
+	logctx.AddStrToContext(r.Context(), LogErrorMessage, err.Error())
+
+	if stack := errs.MarshalStack(err); stack != nil {
+		logctx.AddToContext(r.Context(), LogStack, stack)
+	}
 
 	var (
-		jsonErr       *json.SyntaxError
-		validationErr *validation.Errors
+		jsonErr        *json.SyntaxError
+		validationErrs *validation.Errors
+		validationErr  validation.Error
 	)
 
 	switch {
 	case errors.Is(err, context.Canceled):
-		http.Error(w, err.Error(), StatusContextCancelled)
+		http.Error(w, "canceled", StatusContextCancelled)
+
 	case errors.Is(err, ErrForbidden):
-		http.Error(w, err.Error(), http.StatusForbidden)
+		HTTPError(w, http.StatusForbidden)
+
 	case errors.Is(err, ErrBasicAuthenticate):
 		w.Header().Set("WWW-Authenticate", "Basic realm=Restricted")
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
+		HTTPError(w, http.StatusUnauthorized)
+
 	case errors.Is(err, ErrUnauthorized):
-		http.Error(w, err.Error(), http.StatusUnauthorized)
+		HTTPError(w, http.StatusUnauthorized)
+
 	case errors.As(err, &jsonErr):
```

Comment:
this jsonErr doesn't seem very useful since it's not exhaustive. It misses several common decode errors. Validation seems sufficient on its own since we will generally wrap serialization errors with some context anyway.

Reply:
I did not touch JSON handling in this PR.  Will review it later.